### DISCUSSION TITLE
Signing releases

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#8843] [Ideas]
Keycloak releases only offers a basic hash (MD5/SHA1) as a way to verify the integrity of releases. MDA/SHA1 have issues, but in addition adds little value other than checking download integrity as the hashes can be generated by anyone, and are hosted on the same site as the downloads themselves.

We should like at options on properly signing releases. Where [sigstore](https://www.sigstore.dev/) seems to be a good candidate.

**Comment by stianst** [comment]
@abstractj @ASzc  FIY

**Comment by ASzc** [comment]
Seems to be based on signed commits in git, and chain of custody thereof. All good stuff I've wanted to implement for a long time, alongside deterministic/reproducible builds. I believe there won't be a negative impact on product, I remember hearing about some downsides to signed commits in git when they were first introduced, but I can't find much on that now, it's possible they worked out the bugs.

**Reply by abstractj** [reply]
@stianst it sounds interesting, and definitely I would like to give it a try. From what I noticed . Not sure if it belongs to the scope of your proposal, but one thing that I would like us to do is to make usage of signed commits (probably overlaps some with sigstore functionality). 

That's possible today with GPG + GitHub https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification. GitHub also provides a way to require commit signing across the organization.

**Comment by SecKatie** [comment]
This is a fantastic idea and [cosign](https://docs.sigstore.dev/cosign/working_with_blobs) would likely be a great tool. If the distribution is packaged into a tar it can be easily distributed using the Github Container Registry and signed.

It might also be valuable to follow some of the guidance [here](https://docs.sigstore.dev/cosign/SBOM) to create and sign an SBOM.

**Comment by tacerus** [comment]
Which public key can I verify the signatures against?

**Reply by bickelj** [reply]
I also am trying to find the Keycloak release-signing key. Where is it?

**Reply by ASzc** [reply]
As far as I'm aware, we haven't progressed to the point of implementing sigstore. There are no signatures to verify

**Reply by tacerus** [reply]
Hi @ASzc, thanks for the input. Every binary download available on the GitHub releases page for Keycloak (https://github.com/keycloak/keycloak/releases) shows an associated PGP signature (the `.asc` files). I would like to verify those against a known and trusted PGP key.

**Reply by bickelj** [reply]
@ASzc We might be talking about two different things.

At https://github.com/keycloak/keycloak/releases/tag/22.0.1 I see [the zip file](https://github.com/keycloak/keycloak/releases/download/22.0.1/keycloak-22.0.1.zip) and [its corresponding `gpg` signature file](https://github.com/keycloak/keycloak/releases/download/22.0.1/keycloak-22.0.1.zip.asc). I was wondering about the public key that generated that signature.

@tacerus You beat me to it!

**Reply by ASzc** [reply]
Ah. I looked into it a little bit, and the release process does indeed use a private key for artifact signing via maven gpg plugin, but what the corresponding public key is I don't know. It's part of encrypted parameters that I don't have access to. https://github.com/keycloak-rel/keycloak-rel/actions/runs/5520870334/job/14947800014

**Reply by bickelj** [reply]
@ASzc Thanks for taking a look and pointing us in the right direction. I too see the key listed at [L77 in that job](https://github.com/keycloak-rel/keycloak-rel/actions/runs/5520870334/workflow#L77) and [L77 in main](https://github.com/keycloak-rel/keycloak-rel/blame/main/.github/workflows/release.yml#L77) but as for the corresponding public key probably the author of [that commit](https://github.com/keycloak-rel/keycloak-rel/commit/18a10815b1fd275cc4650fd2ff5e530d3cc1a65a) would know, perhaps a very busy person, @stianst!

**Reply by bickelj** [reply]
@tacerus I found a key `439DEBB700FFCCC0F0CDBE57FDD0432729B196C0` at https://keyserver.ubuntu.com/pks/lookup?search=439DEBB700FFCCC0F0CDBE57FDD0432729B196C0&fingerprint=on&op=index actually, leading to https://keyserver.ubuntu.com/pks/lookup?op=get&search=0xfdd0432729b196c0 and ultimately:
```
-----BEGIN PGP PUBLIC KEY BLOCK-----
Comment: Hostname: 
Version: Hockeypuck 2.1.0-222-g25248d4

xjMEYgZZyBYJKwYBBAHaRw8BAQdA5SWUI7Cr6eOiP5j7cOPm/cRxUCAB1F3v4VoH
8ymdcDzNG0tleWNsb2FrIDxib3RAa2V5Y2xvYWsub3JnPsKaBBMWCgBCFiEEQ53r
twD/zMDwzb5X/dBDJymxlsAFAmIGWcgCGwMFCQPCZwAFCwkIBwIDIgIBBhUKCQgL
AgQWAgMBAh4HAheAAAoJEP3QQycpsZbAxiwA/jgODpx5h5iUZTAZClqWIp7svpWP
RP8SCAAQiET7193PAQDYV8J1DJLVEwDLJsyioAu3tWf7zhd9HKCgtjYTRiMXA844
BGIGWcgSCisGAQQBl1UBBQEBB0BvOcb6PgHENBenKTaWBhAb8+fWo+fv9K64U6mf
iLFAVwMBCAfCfgQYFgoAJhYhBEOd67cA/8zA8M2+V/3QQycpsZbABQJiBlnIAhsM
BQkDwmcAAAoJEP3QQycpsZbAvD4A/jeoJR0XnNQ23G1dB616SNan/1SYeqDocS5h
ZyjdIyXRAQD6nfooGguhtFCkaHVqWdwNB3SPslHJWrKuKNmXoRZ7AA==
=VZ1W
-----END PGP PUBLIC KEY BLOCK-----
```

Whether this is the expected key is still a question, but this is a start.

**Reply by tacerus** [reply]
Thanks, this is a start. However I would personally not trust it until confirmed by project maintainers (and ideally have the fingerprint published in the repository or website).

**Reply by bickelj** [reply]
The above one is expired now. A similar path (https://keyserver.ubuntu.com/pks/lookup?search=861AB50E8CC6611FB6BC01A6B8F12EA26FD6EEBA&fingerprint=on&op=index) led to this:
```
-----BEGIN PGP PUBLIC KEY BLOCK-----
Comment: Hostname: 
Version: Hockeypuck 2.2

xjMEZcsYGxYJKwYBBAHaRw8BAQdAA32swlQ3Z3tyI0iJAY0XMQEVwUXe7nHnjKWs
OCdIPNjNJUtleWNsb2FrIEJvdCA8a2V5Y2xvYWsuYm90QGdtYWlsLmNvbT7CmQQT
FgoAQRYhBIYatQ6MxmEftrwBprjxLqJv1u66BQJlyxgbAhsDBQkFo5qABQsJCAcC
AiICBhUKCQgLAgQWAgMBAh4HAheAAAoJELjxLqJv1u66mOwBANRuVTHzqTknxFKQ
pi5Mrb2oBbImTEgLkiUYzKidRMlxAP9C01g/l1/6EqGoW8EVwx5ZpjvoyPa9xGrH
oYk8lSh8Ds44BGXLGBsSCisGAQQBl1UBBQEBB0BP0g9y7SpLR8mx+JvOKLactT1O
ukxvjuc25/TqxVKDJQMBCAfCfgQYFgoAJhYhBIYatQ6MxmEftrwBprjxLqJv1u66
BQJlyxgbAhsMBQkFo5qAAAoJELjxLqJv1u66L0sBANSOIz9YtRp6lKeUf/MFp7aX
F3SbVD+A1zcxAkG7zb8UAP9wEZEe9LJRj+vvMA//B2dkldazB5XSd8kQMzbCZHwL
Cw==
=IJm3
-----END PGP PUBLIC KEY BLOCK-----
```

**Comment by stianst** [comment]
The keys we are currently using are now published on the website; https://www.keycloak.org/downloads - https://www.keycloak.org/keys
